eks/sync.go

```diff
@@ -0,0 +1,351 @@
+	}
+	if shouldWait {
+		logger.Info("Waiting until new image for coredns is rolled out.")
+		// Ideally we will implement the following routine using the raw client-go library, but implementing this
+		// functionality directly on the API is fairly complex, and thus we rely on the built in mechanism in kubectl
+		// instead.
+		args := []string{
+			"rollout",
+			"status",
+			fmt.Sprintf("deployment/%s", corednsDeploymentName),
+			"-n", componentNamespace,
+			"--timeout", waitTimeout,
+		}
+		return kubectl.RunKubectl(kubectlOptions, args...)
+	}
+	return nil
+}
+
+// getCurrentDeployedCoreDNSImage will return the currently configured coredns image on the deployment.
+func getCurrentDeployedCoreDNSImage(clientset *kubernetes.Clientset) (string, error) {
+	deployment, err := clientset.AppsV1().Deployments(componentNamespace).Get(context.Background(), corednsDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	deploymentContainers := deployment.Spec.Template.Spec.Containers
+	if len(deploymentContainers) != 1 {
+		err := CoreComponentUnexpectedConfigurationErr{
+			component: "coredns",
+			reason:    fmt.Sprintf("unexpected number of containers (%d)", len(deploymentContainers)),
+		}
+		return "", errors.WithStackTrace(err)
+	}
+	return deploymentContainers[0].Image, nil
+}
+
+// updateCoreDNSDeploymentImage will update the deployed coredns Deployment to the specified target container image.
+func updateCoreDNSDeploymentImage(clientset *kubernetes.Clientset, targetImage string) error {
+	patch := []jsonpatch.PatchString{
+		{
+			Op: jsonpatch.ReplaceOp,
+			// Patch the first container's image field
+			Path:  "/spec/template/spec/containers/0/image",
+			Value: targetImage,
+		},
+	}
+	patchOpJson, err := json.Marshal(patch)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	deploymentAPI := clientset.AppsV1().Deployments(componentNamespace)
+	if _, err := deploymentAPI.Patch(context.Background(), corednsDeploymentName, k8stypes.JSONPatchType, patchOpJson, metav1.PatchOptions{}); err != nil {
+		return errors.WithStackTrace(err)
+	}
+	return nil
+}
+
+// updateVPCCNI will apply the manifest to deploy the latest patch release of the target AWS VPC CNI version. Ideally we
+// would implement this using the raw Kubernetes API, but the CNI manifest contains additional resources on top of the
+// daemonset, and thus it is better to apply the manifests directly using kubectl than to translate it into underlying
+// API calls.
+func updateVPCCNI(kubectlOptions *kubectl.KubectlOptions, region string, vpcCNIVersion string) error {
+	var manifestPath string
+
+	// Figure out the manifest URL based on region
+	// Reference: https://docs.aws.amazon.com/eks/latest/userguide/update-cluster.html
+	if strings.HasPrefix(region, "cn-") {
+		manifestPath = fmt.Sprintf("https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/release-%s/config/v%s/aws-k8s-cni-cn.yaml", vpcCNIVersion, vpcCNIVersion)
```

Comment:
NIT: It might be slightly more DRY and readable to `Sprintf` the URL once, like:

```
awsK8sCNIURL := fmt.Sprintf("https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/release-%s/config/v%s/", vpcCNIVersion, vpcCNIVersion)
if strings.HasPrefix(region, "cn-") {
    manifestPath = awsK8sCNIURL + "aws-k8s-cni-cn.yaml"
} else if region == "us-gov-east-1" {
... etc etc ...
}
```

Reply:
Ah good point! Updated: 87ea8c4